bundle/edu.gemini.horizons.server/src/main/scala/edu/gemini/horizons/server/backend/HS2.scala

```diff
@@ -0,0 +1,299 @@
+  type HS2[A] = EitherT[IO, HS2Error, A]
+  object HS2 {
+    def fromDisjunction[A](a: HS2Error \/ A): HS2[A] = EitherT(IO(a))
+    def delay[A](a: => A): HS2[A] = EitherT(IO(a.right))
+  }
+
+  /** The type of failures that might arise when talking to HORIZONS. */
+  sealed trait HS2Error
+  case class  HorizonsError(e: HorizonsException) extends HS2Error
+  case class  ParseError(input: List[String], message: String) extends HS2Error
+  case object EphemerisEmpty extends HS2Error
+
+  /** A value annotated with a name. Used when multiple results are returned. */
+  case class Row[A](a: A, name: String)
+
+  /** A partial-name search specification for HORIZONS. */
+  sealed abstract class Search[A](val queryString: String) extends Product with Serializable
+  object Search {
+    final case class Comet(partial: String)     extends Search[HorizonsDesignation.Comet](s"COMNAM=$partial*;CAP")
+    final case class Asteroid(partial: String)  extends Search[HorizonsDesignation.Asteroid](s"ASTNAM=$partial*")
+    final case class MajorBody(partial: String) extends Search[HorizonsDesignation.MajorBody](s"$partial")
+  }
+
+  /** Construct a program that performs a partial-name search. */
+  def search[A](search: Search[A]): HS2[List[Row[A]]] = {
+
+    val queryParams: Map[String, String] =
+      Map(
+        BATCH      -> "1",
+        TABLE_TYPE -> OBSERVER_TABLE,
+        CSV_FORMAT -> NO,
+        EPHEMERIS  -> NO,
+        COMMAND    -> s"'${search.queryString}'"
+      )
+
+    // Lift our pure String \/ List[Row[A]] into HS2
+    def parseLines(lines: List[String]): HS2[List[Row[A]]] =
+      HS2.fromDisjunction(parseResponse(search, lines).leftMap(ParseError(lines, _)))
+
+    // And finally
+    horizonsRequestLines(queryParams) >>= parseLines
+
+  }
+
+  /** Convenience method; looks up ephemeris for the current semester. */
+  def lookupEphemeris(target: HorizonsDesignation, site: Site, elems: Int): HS2[Ephemeris] =
+    HS2.delay(new Semester(site)).flatMap(lookupEphemeris(target, site, elems, _))
+
+  /** Convenience method; looks up ephemeris for the given semester. */
+  def lookupEphemeris(target: HorizonsDesignation, site: Site, elems: Int, sem: Semester): HS2[Ephemeris] =
+    lookupEphemeris(target, site, sem.getStartDate(site), sem.getEndDate(site), elems)
+
+  /** 
+   * Look up the ephemeris for the given target when viewed from the given site, requesting `elems`
+   * total elements spread uniformly over the over the given time period. The computed ephemeris may
+   * contain slightly more or fewer entries than requested due to rounding, or many fewer for very
+   * short timespans (no more than one entry will be returned per minute).
+   */
+  def lookupEphemeris(target: HorizonsDesignation, site: Site, start: Date, stop: Date, elems: Int): HS2[Ephemeris] = {
+
+    def formatDate(date: Date): String = {
+      val df = new SimpleDateFormat("yyyy-MMM-dd HH:mm:ss")
+      s"'${df.format(date)}'"
+    }
```

Comment:
These things are [straightforward to build](https://github.com/tpolecat/doobie/blob/master/core/src/main/scala/doobie/util/catchable.scala) in terms of `Catchable` (which is the exception-trapping effect that `Task` implements) but nobody has done this for scalaz because the plan was to replace both `Task` and `IO` with something better by now. I might do it for 7.2 if nothing happens in the interim; the library support for `IO` is much better, but it doesn't support concurrency (which is why everyone is using `Task` instead).